sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala

```diff
@@ -0,0 +1,292 @@
+   * list of `inType`.
+   */
+  private def implicitCast(inType: DataType, expectedType: AbstractDataType): Option[DataType] = {
+    (inType, expectedType) match {
+      // If the expected type equals the input type, no need to cast.
+      case _ if expectedType.acceptsType(inType) => Some(inType)
+
+      // Cast null type (usually from null literals) into target types
+      case (NullType, target) => Some(target.defaultConcreteType)
+
+      // If input is a numeric type but not decimal, and we expect a decimal type,
+      // cast the input to decimal.
+      case (d: NumericType, DecimalType) => Some(DecimalType.forType(d))
+
+      case (n1: NumericType, n2: NumericType) =>
+        val widerType = findWiderTypeForTwo(n1, n2)
+        widerType match {
+          // if the expected type is Float type, we should still return Float type.
+          case Some(DoubleType) if n1 != DoubleType && n2 == FloatType => Some(FloatType)
+
+          case Some(dt) if dt == n2 => Some(dt)
+
+          case _ => None
+        }
+
+      case (DateType, TimestampType) => Some(TimestampType)
+
+      // When we reach here, input type is not acceptable for any types in this type collection,
+      // try to find the first one we can implicitly cast.
+      case (_, TypeCollection(types)) =>
+        types.flatMap(implicitCast(inType, _)).headOption
+
+      // Implicit cast between array types.
+      //
+      // Compare the nullabilities of the from type and the to type, check whether the cast of
+      // the nullability is resolvable by the following rules:
+      // 1. If the nullability of the to type is true, the cast is always allowed;
+      // 2. If the nullabilities of both the from type and the to type are false, the cast is
+      //    allowed.
+      // 3. Otherwise, the cast is not allowed
+      case (ArrayType(fromType, containsNullFrom), ArrayType(toType: DataType, containsNullTo))
+          if Cast.resolvableNullability(containsNullFrom, containsNullTo) =>
+        implicitCast(fromType, toType).map(ArrayType(_, containsNullTo))
+
+      // Implicit cast between Map types.
+      // Follows the same semantics of implicit casting between two array types.
+      // Refer to documentation above.
+      case (MapType(fromKeyType, fromValueType, fn), MapType(toKeyType, toValueType, tn))
+          if Cast.resolvableNullability(fn, tn) =>
+        val newKeyType = implicitCast(fromKeyType, toKeyType)
+        val newValueType = implicitCast(fromValueType, toValueType)
+        if (newKeyType.isDefined && newValueType.isDefined) {
+          Some(MapType(newKeyType.get, newValueType.get, tn))
+        } else {
+          None
+        }
+
+      case _ => None
+    }
+  }
+
+  override def canCast(from: DataType, to: DataType): Boolean = AnsiCast.canCast(from, to)
+
+  /**
+   * Promotes string literals that appear in arithmetic and comparison expressions.
+   */
+  object PromoteStringLiterals extends TypeCoercionRule {
```

Comment:
Could you add more tests for this new rule?

Reply:
Yes, will do